client/src/entrypoints/admin/bulk-actions.js

```diff
@@ -0,0 +1,165 @@
+function SelectBulkActionsFilter(e) {
+  const changeEvent = new Event('change');
+  for (const el of document.querySelectorAll(`.${BULK_ACTION_PAGE_CHECKBOX_INPUT}`)) {
+    if (el.checked === e.target.checked) continue;
+    el.checked = e.target.checked;
+    el.dispatchEvent(changeEvent);
+  }
+}
+
+
+/* Event listener for individual page checkbox */
+function SelectBulkActionsCheckboxes(e) {
+  if (checkedState.selectAllInListing) checkedState.selectAllInListing = false;
+  const prevLength = checkedState.checkedPages.size;
+  if (e.target.checked) checkedState.checkedPages.add(+e.target.dataset.pageId);
+  else {
+    /* unchecks `Select all` checkbox as soon as one page is unchecked */
+    document.querySelector(`.${BULK_ACTION_SELECT_ALL_CHECKBOX_TH} input`).checked = false;
+    checkedState.checkedPages.delete(+e.target.dataset.pageId);
+  }
+
+  if (checkedState.checkedPages.size === 0) {
+    /* when all checboxes are unchecked */
+    document.querySelectorAll(`.${TABLE_HEADERS_TR} > th`).forEach(el => el.classList.remove('u-hidden'));
+    document.querySelector(`.${BULK_ACTION_CHOICES_DIV}`).classList.add('u-hidden');
+    document.querySelectorAll(`.${BULK_ACTION_PAGE_CHECKBOX_INPUT}`).forEach(el => el.classList.remove('show'));
+    document.querySelector(`.${BULK_ACTION_SELECT_ALL_CHECKBOX_TH}`).setAttribute('colspan', '1');
+  } else if (checkedState.checkedPages.size === 1 && prevLength === 0) {
+    /* when 1 checkbox is checked for the first time */
+    document.querySelectorAll(`.${BULK_ACTION_PAGE_CHECKBOX_INPUT}`).forEach(el => {
+      el.classList.remove('show');
+      el.classList.add('show');
+    });
+    document.querySelectorAll(`.${TABLE_HEADERS_TR} > th`).forEach(el => el.classList.add('u-hidden'));
+    document.querySelector(`.${BULK_ACTION_SELECT_ALL_CHECKBOX_TH}`).classList.remove('u-hidden');
+    document.querySelector(`.${BULK_ACTION_CHOICES_DIV}`).classList.remove('u-hidden');
+    document.querySelector(`.${BULK_ACTION_SELECT_ALL_CHECKBOX_TH}`).setAttribute('colspan', '6');
+  }
+
+  if (checkedState.checkedPages.size === checkedState.numPages) {
+    /* when all checkboxes in the page are checked */
+    document.querySelector(`.${BULK_ACTION_SELECT_ALL_CHECKBOX_TH} input`).checked = true;
+  }
+
+  if (checkedState.checkedPages.size > 0) {
+    /* Update text on number of pages */
+    let numPagesSelected = '';
+    if (checkedState.checkedPages.size === checkedState.numPages) {
+      document.querySelector(`.${BULK_ACTION_NUM_PAGES_IN_LISTING}`).classList.remove('u-hidden');
+      if (checkedState.numPages === 1) {
+        numPagesSelected = wagtailConfig.STRINGS.NUM_PAGES_SELECTED_SINGULAR;
+      } else {
+        numPagesSelected = wagtailConfig.STRINGS.NUM_PAGES_SELECTED_ALL.replace('{0}', checkedState.checkedPages.size);
+      }
+    } else {
+      document.querySelector(`.${BULK_ACTION_NUM_PAGES_IN_LISTING}`).classList.add('u-hidden');
+      if (checkedState.checkedPages.size === 1) {
```

Comment:
The `numPagesSelected` logic seems to be repeated twice. I would suggest refactoring to only find it once.

Reply:
@KIRA009 just tagging you because this might have been overlooked from a while back